src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -1627,25 +1702,60 @@ public void forEachRemaining(Consumer<? super Map.Entry<K, V>> action) {
                                 @Override
                                 public void accept(Entry<K, StableValueImpl<V>> inner) {
                                     final K k = inner.getKey();
-                                    action.accept(new NullableKeyValueHolder<>(k, inner.getValue().orElseSet(new Supplier<V>() {
-                                        @Override public V get() { return mapper.apply(k); }})));
+                                    action.accept(new StableEntry<>(k, inner.getValue(), new Supplier<V>() {
+                                        @Override public V get() { return outer.outer.mapper.apply(k); }}));
                                 }
                             };
                     delegateIterator.forEachRemaining(innerAction);
                 }
+
+                // For @ValueBased
+                private static  <K, V> LazyMapIterator<K, V> of(StableMapEntrySet<K, V> outer) {
+                    return new LazyMapIterator<>(outer);
+                }
+
             }
         }
 
+        private record StableEntry<K, V>(K getKey, // trick
+                                         StableValueImpl<V> stableValue,
+                                         Supplier<? extends V> supplier) implements Map.Entry<K, V> {
+
+            @Override public V setValue(V value) { throw uoe(); }
+            @Override public V getValue() { return stableValue.orElseSet(supplier); }
+            @Override public int hashCode() { return hash(getKey()) ^ hash(getValue()); }
+            @Override public String toString() { return getKey() + "=" + stableValue.toString(); }
+            @Override public boolean equals(Object o) {
+                return o instanceof Map.Entry<?, ?> e
+                        && Objects.equals(getKey(), e.getKey())
+                        // Invoke `getValue()` as late as possible to avoid evaluation
+                        && Objects.equals(getValue(), e.getValue());
```

Comment:
```suggestion
                        && Objects.equals(e.getValue(), getValue());
```
You already integrated and it's probably not that important, but this would additionally avoid evaluation if `e.getValue()` throws.

Reply:
This is in the realms of microoptimization now...